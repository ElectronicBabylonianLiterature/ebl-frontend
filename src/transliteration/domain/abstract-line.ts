import { Token } from './token'
import { immerable } from 'immer'

export interface LineBaseDto {
  readonly type: string
  readonly prefix: string
  readonly content: readonly Token[]
}

export interface DollarAndAtLineDto extends LineBaseDto {
  readonly displayValue: string
}

export abstract class AbstractLine {
  readonly [immerable] = true
  abstract get type(): string
  readonly prefix: string
  readonly content: readonly Token[]

  constructor(prefix: string, content: readonly Token[]) {
    this.prefix = prefix
    this.content = content
  }
}

export type Ruling = 'SINGLE' | 'DOUBLE' | 'TRIPLE'
