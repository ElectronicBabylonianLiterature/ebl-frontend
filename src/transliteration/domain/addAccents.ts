import {
  AkkadianWord,
  NamedSign,
  Token,
  ValueToken,
} from 'transliteration/domain/token'

const vowels: ReadonlySet<string> = new Set([
  'a',
  'e',
  'i',
  'u',
  'A',
  'E',
  'I',
  'U',
])
const acuteAccents: ReadonlyMap<string, string> = new Map([
  ['A', '\u00C1'],
  ['E', '\u00C9'],
  ['I', '\u00CD'],
  ['O', '\u00D3'],
  ['U', '\u00DA'],
  ['a', '\u00E1'],
  ['e', '\u00E9'],
  ['i', '\u00ED'],
  ['o', '\u00F3'],
  ['u', '\u00FA'],
])
const graveAccents: ReadonlyMap<string, string> = new Map([
  ['A', '\u00C0'],
  ['E', '\u00C8'],
  ['I', '\u00CC'],
  ['O', '\u00D2'],
  ['U', '\u00D9'],
  ['a', '\u00E0'],
  ['e', '\u00E8'],
  ['i', '\u00EC'],
  ['o', '\u00F2'],
  ['u', '\u00F9'],
])
const h = 'h'
const hBreve = '\u1E2B'

function isValueToken(token: Token): token is ValueToken {
  return token.type === 'ValueToken'
}

function isVowel(letter: string): boolean {
  return vowels.has(letter)
}

function addGraveAccent(letter: string): string {
  return graveAccents.get(letter) ?? letter
}

function addAcuteAccent(letter: string): string {
  return acuteAccents.get(letter) ?? letter
}

class Accumulator {
  private isSubIndexConverted = false
  private isFirstWovel = true
  private tokens: Token[] = []

  constructor(private subIndex: number | null | undefined) {}

  get result(): [Token[], boolean] {
    return [this.tokens, this.isSubIndexConverted]
  }

  addToken(token: Token): Accumulator {
    this.tokens.push(
      isValueToken(token)
        ? {
            ...token,
            value: this.addAccentsToValue(token.value),
          }
        : token
    )
    return this
  }

  private addAccentsToValue(value: string): string {
    return value
      .split('')
      .map((letter) => this.visitLetter(letter))
      .join('')
  }

  private visitLetter(letter: string): string {
    if (this.isFirstWovel && isVowel(letter)) {
      return this.visitVowel(letter)
    } else if (letter === h) {
      return hBreve
    } else {
      return letter
    }
  }

  private visitVowel(vowel: string): string {
    if (this.subIndex === 2) {
      this.isFirstWovel = false
      this.isSubIndexConverted = true
      return addAcuteAccent(vowel)
    } else if (this.subIndex === 3) {
      this.isFirstWovel = false
      this.isSubIndexConverted = true
      return addGraveAccent(vowel)
    } else {
      return vowel
    }
  }
}

export function addAccents(
  namedSign: NamedSign
): readonly [readonly Token[], boolean] {
  return namedSign.nameParts.reduce(
    (acc, token) => acc.addToken(token),
    new Accumulator(namedSign.subIndex)
  ).result
}

export function addBreves(word: AkkadianWord): AkkadianWord {
  return {
    ...word,
    parts: word.parts.map((part) =>
      isValueToken(part)
        ? {
            ...part,
            value: part.value.replace(h, hBreve),
          }
        : part
    ),
  }
}
