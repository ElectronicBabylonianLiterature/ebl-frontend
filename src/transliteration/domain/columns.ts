import _ from 'lodash'
import { Token } from './token'
import { isAkkadianWord, isColumn } from './type-guards'
import {
  AnnotationLineAccumulator,
  LineAccumulator,
} from 'transliteration/ui/LineAccumulator'
import { PhoneticProps } from 'akkadian/application/phonetics/segments'

export interface TextLineColumn {
  span: number | null
  content: Token[]
}

function updatePhoneticPropsContext(
  content: Token[],
  index: number,
  phoneticProps?: PhoneticProps
): PhoneticProps {
  const previousWord = _.find(content.slice(0, index).reverse(), (token) =>
    isAkkadianWord(token)
  )
  const nextWord = _.find(content.slice(index + 1), (token) =>
    isAkkadianWord(token)
  )
  return {
    ...phoneticProps,
    wordContext: {
      ...(previousWord &&
        isAkkadianWord(previousWord) && {
          previousWord,
        }),
      ...(nextWord &&
        isAkkadianWord(nextWord) && {
          nextWord,
        }),
    },
  }
}

export const defaultSpan = 1

export function lineAccFromColumns({
  columns,
  isInLineGroup = false,
  showMeter = false,
  showIpa = false,
  phoneticProps,
  highlightLemmas = [],
  isInPopover,
}: {
  columns: readonly TextLineColumn[]
  isInLineGroup?: boolean
  showMeter?: boolean
  showIpa?: boolean
  phoneticProps?: PhoneticProps
  highlightLemmas: readonly string[]
  isInPopover?: boolean
}): LineAccumulator {
  return columns.reduce((acc: LineAccumulator, column) => {
    acc.addColumn(column.span)
    column.content.reduce(
      (acc: LineAccumulator, token: Token, index: number) => {
        acc.addColumnToken(
          token,
          index,
          updatePhoneticPropsContext(column.content, index, phoneticProps),
          _.isEmpty(_.intersection(token.uniqueLemma, highlightLemmas))
            ? []
            : ['highlight'],
          isInPopover
        )
        return acc
      },
      acc
    )
    return acc
  }, new LineAccumulator(isInLineGroup, showMeter, showIpa))
}

export function annotationLineAccFromColumns(
  columns: readonly TextLineColumn[]
): AnnotationLineAccumulator {
  return columns.reduce((acc: AnnotationLineAccumulator, column) => {
    acc.addColumn(column.span)
    column.content.reduce(
      (acc: AnnotationLineAccumulator, token: Token, index: number) => {
        acc.addColumnToken(token, index)
        return acc
      },
      acc
    )
    return acc
  }, new AnnotationLineAccumulator())
}

export function numberOfColumns(columns: readonly TextLineColumn[]): number {
  return _(columns)
    .map((column) => column.span ?? defaultSpan)
    .sum()
}

export function maxColumns(lines: readonly TextLineColumn[][]): number {
  return _(lines).map(numberOfColumns).max() ?? 1
}

export function containsColumns(tokens: readonly Token[]): boolean {
  return tokens.some(isColumn)
}

export function createColumns(tokens: readonly Token[]): TextLineColumn[] {
  const hasColumns = containsColumns(tokens)
  return _.reduce<Token, TextLineColumn[]>(
    tokens,
    (columns, current) => {
      if (isColumn(current)) {
        if (_.isEmpty(columns) && current.number === null) {
          columns.push({
            span: defaultSpan,
            content: [],
          })
        }
        columns.push({
          span: current.number ?? defaultSpan,
          content: [],
        })
      } else if (_.isEmpty(columns)) {
        columns.push({
          span: hasColumns ? 1 : null,
          content: [current],
        })
      } else {
        _.last(columns)?.content.push(current)
      }
      return columns
    },
    []
  )
}
