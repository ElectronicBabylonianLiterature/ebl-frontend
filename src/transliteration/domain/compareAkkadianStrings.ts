import _ from 'lodash'
import alphabet from './alphabet.json'

function removeIgnoredCharacters(word: string): string {
  return word.replace(/\]|\?|\[|-|\]|\./g, '')
}

function checkForInvalidCharacters(...words: string[]): void {
  const invalidCharacters = words
    .flatMap((word) => word.split(''))
    .filter((character) => !alphabet.includes(character))
  if (!_.isEmpty(invalidCharacters)) {
    throw new Error(`Invalid character(s) ${invalidCharacters} in the input`)
  }
}

function compareAlphabet(word: string, anotherWord: string): number {
  return alphabet.indexOf(word) - alphabet.indexOf(anotherWord)
}

export default function compareAkkadianStrings(
  word: string,
  anotherWord: string
): number {
  const replacedWord = removeIgnoredCharacters(word)
  const anotherWordReplaced = removeIgnoredCharacters(anotherWord)
  checkForInvalidCharacters(replacedWord, anotherWordReplaced)

  return (
    _(replacedWord)
      .split('')
      .zipWith(anotherWordReplaced.split(''), compareAlphabet)
      .filter((result) => result !== 0)
      .map(Math.sign)
      .head() ?? Math.sign(replacedWord.length - anotherWordReplaced.length)
  )
}
