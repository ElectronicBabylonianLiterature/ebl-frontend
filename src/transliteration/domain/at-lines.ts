import { ColumnLabel, SurfaceLabel, ObjectLabel } from './labels'
import { AbstractLine, DollarAndAtLineDto } from './abstract-line'

export abstract class AtLine extends AbstractLine {
  readonly displayValue: string

  constructor(data: DollarAndAtLineDto) {
    super('@', data.content)
    this.displayValue = `(${data.displayValue})`
  }
}

export interface SealAtLineDto extends DollarAndAtLineDto {
  readonly type: 'SealAtLine'
  readonly number: number
}

export class SealAtLine extends AtLine {
  readonly type = 'SealAtLine'
  readonly number: number

  constructor(data: SealAtLineDto) {
    super(data)
    this.number = data.number
  }
}

export interface HeadingAtLineDto extends DollarAndAtLineDto {
  readonly type: 'HeadingAtLine'
  readonly number: number
}

export class HeadingAtLine extends AtLine {
  readonly type = 'HeadingAtLine'
  readonly number: number

  constructor(data: HeadingAtLineDto) {
    super(data)
    this.number = data.number
  }
}

export interface ColumnAtLineDto extends DollarAndAtLineDto {
  readonly type: 'ColumnAtLine'
  readonly column_label: ColumnLabel
}

export class ColumnAtLine extends AtLine {
  readonly type = 'ColumnAtLine'
  readonly label: ColumnLabel

  constructor(data: ColumnAtLineDto) {
    super(data)
    this.label = data.column_label
  }
}

export interface DiscourseAtLineDto extends DollarAndAtLineDto {
  readonly type: 'DiscourseAtLine'
  readonly discourse_label: string
}

export class DiscourseAtLine extends AtLine {
  readonly type = 'DiscourseAtLine'
  readonly label: string

  constructor(data: DiscourseAtLineDto) {
    super(data)
    this.label = data.discourse_label
  }
}

export interface SurfaceAtLineDto extends DollarAndAtLineDto {
  readonly type: 'SurfaceAtLine'
  readonly surface_label: SurfaceLabel
}

export class SurfaceAtLine extends AtLine {
  readonly type = 'SurfaceAtLine'
  readonly label: SurfaceLabel

  constructor(data: SurfaceAtLineDto) {
    super(data)
    this.label = data.surface_label
  }
}

export interface ObjectAtLineDto extends DollarAndAtLineDto {
  readonly type: 'ObjectAtLine'
  readonly label: ObjectLabel
}

export class ObjectAtLine extends AtLine {
  readonly type = 'ObjectAtLine'
  readonly label: ObjectLabel

  constructor(data: ObjectAtLineDto) {
    super(data)
    this.label = data.label
  }
}

export interface DivisionAtLineDto extends DollarAndAtLineDto {
  readonly type: 'DivisionAtLine'
  readonly number: number | null
  readonly text: string
}

export class DivisionAtLine extends AtLine {
  readonly type = 'DivisionAtLine'
  readonly number: number | null
  readonly text: string

  constructor(data: DivisionAtLineDto) {
    super(data)
    this.number = data.number
    this.text = data.text
  }
}

export interface CompositeAtLineDto extends DollarAndAtLineDto {
  readonly type: 'CompositeAtLine'
  readonly composite: string
  readonly number: number | null
  readonly text: string
}

export class CompositeAtLine extends AtLine {
  readonly type = 'CompositeAtLine'
  readonly composite: string
  readonly number: number | null
  readonly text: string

  constructor(data: CompositeAtLineDto) {
    super(data)
    this.composite = data.composite
    this.number = data.number
    this.text = data.text
  }
}
