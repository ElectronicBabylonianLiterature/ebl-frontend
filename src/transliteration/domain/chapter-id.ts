import { TextId } from 'transliteration/domain/text-id'
import isEmpty from 'lodash/fp/isEmpty'
import join from 'lodash/fp/join'
import reject from 'lodash/fp/reject'
import flow from 'lodash/fp/flow'

export const defaultName = '-'

export interface ChapterId {
  readonly textId: TextId
  readonly stage: string
  readonly name: string
}

export function chapterIdToString(id: ChapterId): string {
  return flow(
    reject(isEmpty),
    join(' ')
  )([id.stage, id.name !== defaultName ? id.name : ''])
}
