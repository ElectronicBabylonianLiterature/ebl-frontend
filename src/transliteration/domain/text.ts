import produce, { immerable, Draft, castDraft } from 'immer'
import _ from 'lodash'
import DictionaryWord from 'dictionary/domain/Word'
import Lemmatization, {
  LemmatizationToken,
  UniqueLemma,
} from 'transliteration/domain/Lemmatization'
import { Line, NoteLine, TextLine } from 'transliteration/domain/line'
import { Word as TransliterationWord } from 'transliteration/domain/token'
import {
  isTextLine,
  isWord,
  isObjectAtLine,
  isSurfaceAtLine,
  isColumnAtLine,
} from 'transliteration/domain/type-guards'
import Lemma from './Lemma'
import { isNoteLine } from './type-guards'
import { LineNumber, LineNumberRange } from './line-number'
import { ObjectLabel, SurfaceLabel, ColumnLabel } from './labels'

export type Notes = ReadonlyMap<number, readonly NoteLine[]>

export function noteNumber(
  notes: Notes,
  lineIndex: number,
  noteIndex: number
): number {
  const numberOfNotesOnPreviousLines = _.sum(
    _.range(0, lineIndex).map((index) => notes.get(index)?.length ?? 0)
  )
  return 1 + noteIndex + numberOfNotesOnPreviousLines
}

export interface GlossaryToken {
  readonly label: Label
  readonly value: string
  readonly word: TransliterationWord
  readonly uniqueLemma: string
  readonly dictionaryWord?: DictionaryWord
}

export class Label {
  readonly [immerable] = true
  readonly object: ObjectLabel | null = null
  readonly surface: SurfaceLabel | null = null
  readonly column: ColumnLabel | null = null
  readonly line: LineNumber | LineNumberRange | null = null

  constructor(
    object: ObjectLabel | null = null,
    surface: SurfaceLabel | null = null,
    column: ColumnLabel | null = null,
    line: LineNumber | LineNumberRange | null = null
  ) {
    this.object = object
    this.surface = surface
    this.column = column
    this.line = line
  }

  setObject(object: ObjectLabel): Label {
    return produce(this, (draft: Draft<Label>) => {
      draft.object = castDraft(object)
    })
  }

  setSurface(surface: SurfaceLabel): Label {
    return produce(this, (draft: Draft<Label>) => {
      draft.surface = castDraft(surface)
    })
  }

  setColumn(column: ColumnLabel): Label {
    return produce(this, (draft: Draft<Label>) => {
      draft.column = castDraft(column)
    })
  }

  setLineNumber(line: LineNumber | LineNumberRange): Label {
    return produce(this, (draft: Draft<Label>) => {
      draft.line = castDraft(line)
    })
  }
}

type LabeledLine = readonly [Label, TextLine]

export class Text {
  readonly allLines: readonly Line[]

  constructor({ lines }: { lines: readonly Line[] }) {
    this.allLines = lines
  }

  get numberOfColumns(): number {
    return (
      _(this.allLines)
        .map((line) => (isTextLine(line) ? line.numberOfColumns : 1))
        .max() ?? 1
    )
  }

  get lines(): readonly Line[] {
    return this.allLines.filter((line) => !isNoteLine(line))
  }

  get notes(): Notes {
    const notes: Map<number, NoteLine[]> = new Map([[0, []]])
    let lineNumber = 0
    for (const line of this.allLines) {
      if (isNoteLine(line)) {
        notes.get(lineNumber)?.push(line)
      } else {
        lineNumber++
        notes.set(lineNumber, [])
      }
    }
    return notes
  }

  get glossary(): [string, readonly GlossaryToken[]][] {
    const [, labeledLines] = _.reduce(
      this.lines,
      (
        [current, lines]: [Label, LabeledLine[]],
        line: Line
      ): [Label, LabeledLine[]] => {
        if (isTextLine(line)) {
          return [
            current,
            [...lines, [current.setLineNumber(line.lineNumber), line]],
          ]
        } else if (isObjectAtLine(line)) {
          return [current.setObject(line.label), lines]
        } else if (isSurfaceAtLine(line)) {
          return [current.setSurface(line.surface_label), lines]
        } else if (isColumnAtLine(line)) {
          return [current.setColumn(line.column_label), lines]
        } else {
          return [current, lines]
        }
      },
      [new Label(), []]
    )

    return _(labeledLines)
      .flatMap(([label, line]) =>
        line.content
          .filter(isWord)
          .filter((token: TransliterationWord) => token.lemmatizable)
          .flatMap(
            (token): GlossaryToken[] =>
              token.uniqueLemma?.map((lemma) => ({
                label: label,
                value: token.value,
                word: token,
                uniqueLemma: lemma,
              })) ?? []
          )
      )
      .groupBy((token) => token.uniqueLemma)
      .toPairs()
      .sortBy(([lemma, tokensByLemma]) => lemma)
      .value()
  }

  createLemmatization(
    lemmas: { [key: string]: Lemma },
    suggestions: { [key: string]: readonly UniqueLemma[] }
  ): Lemmatization {
    return new Lemmatization(
      this.allLines.map((line) => line.prefix),
      this.allLines
        .map((line) => line.content)
        .map((tokens) =>
          tokens.map((token) =>
            token.lemmatizable
              ? new LemmatizationToken(
                  token.value,
                  true,
                  (token.uniqueLemma || []).map((id) => lemmas[id]),
                  suggestions[token.cleanValue]
                )
              : new LemmatizationToken(token.value, false)
          )
        )
    )
  }
}
Text[immerable] = true
