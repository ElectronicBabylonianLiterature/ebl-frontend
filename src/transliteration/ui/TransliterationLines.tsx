import _ from 'lodash'
import React, { FunctionComponent, PropsWithChildren } from 'react'
import { Text, Notes } from 'transliteration/domain/text'
import { NoteLinks, createLineId } from './note-links'
import DisplayRulingDollarLine from './rulings'
import DisplayTextLine from './text-line'
import { DisplayDollarAndAtLine } from './dollar-and-at-lines'
import { LineProps } from './LineProps'
import { AbstractLine } from 'transliteration/domain/abstract-line'
import { defaultLabels, Labels } from 'transliteration/domain/labels'
import {
  isColumnAtLine,
  isObjectAtLine,
  isSurfaceAtLine,
  isTranslationLine,
} from 'transliteration/domain/type-guards'
import DisplayTranslationLine from './DisplayTranslationLine'
import DisplayControlLine from './DisplayControlLine'
import { DisplayParallelLine } from './parallel-line'
import TranslationColumn from 'transliteration/ui/TranslationColumn'

export type LineComponentMap = ReadonlyMap<string, FunctionComponent<LineProps>>
interface TextDisplayProps {
  text: Text
  activeLine?: string
  language: string | null
}

export const defaultLineComponents: LineComponentMap = new Map([
  ['TextLine', DisplayTextLine],
  ['RulingDollarLine', DisplayRulingDollarLine],
  ['LooseDollarLine', DisplayDollarAndAtLine],
  ['ImageDollarLine', DisplayDollarAndAtLine],
  ['SealDollarLine', DisplayDollarAndAtLine],
  ['StateDollarLine', DisplayDollarAndAtLine],
  ['SealAtLine', DisplayDollarAndAtLine],
  ['ColumnAtLine', DisplayDollarAndAtLine],
  ['HeadingAtLine', DisplayDollarAndAtLine],
  ['DiscourseAtLine', DisplayDollarAndAtLine],
  ['SurfaceAtLine', DisplayDollarAndAtLine],
  ['ObjectAtLine', DisplayDollarAndAtLine],
  ['DivisionAtLine', DisplayDollarAndAtLine],
  ['CompositeAtLine', DisplayDollarAndAtLine],
  ['TranslationLine', DisplayTranslationLine],
  ['ParallelFragment', DisplayParallelLine],
  ['ParallelText', DisplayParallelLine],
  ['ParallelComposition', DisplayParallelLine],
])

function FirstLineNotes({
  notes,
  columns,
}: {
  notes: Notes
  columns: number
}): JSX.Element {
  const hasNotes = !_.isEmpty(notes.get(0))
  return (
    <>
      {hasNotes && (
        <tr id={createLineId(0)}>
          <td></td>
          <td colSpan={columns}></td>
          <td>
            <NoteLinks notes={notes} lineNumber={0} />
          </td>
        </tr>
      )}
    </>
  )
}

export function getCurrentLabels(labels: Labels, line: AbstractLine): Labels {
  if (isObjectAtLine(line)) {
    return { ...labels, object: line.label }
  } else if (isSurfaceAtLine(line)) {
    return { ...labels, surface: line.label }
  } else if (isColumnAtLine(line)) {
    return { ...labels, column: line.label }
  } else {
    return labels
  }
}

export type TranslationStyle = 'inline' | 'standoff'

function DisplayRow({
  line,
  lineIndex,
  columns,
  labels,
  activeLine,
  notes,
  children,
}: PropsWithChildren<
  LineProps & {
    lineIndex: number
    notes: Notes
  }
>): JSX.Element {
  const LineComponent =
    defaultLineComponents.get(line.type) || DisplayControlLine
  const lineNumber = lineIndex + 1
  return (
    <tr id={createLineId(lineNumber)}>
      <LineComponent
        line={line}
        lineIndex={lineIndex}
        columns={columns}
        labels={labels}
        activeLine={activeLine}
      />
      <td>
        <NoteLinks notes={notes} lineNumber={lineNumber} />
      </td>
      {children}
    </tr>
  )
}

function skipLine(line: AbstractLine, language: string | null): boolean {
  return !!language && isTranslationLine(line) && line.language !== language
}

function DisplaySingleColumnText({
  text,
  activeLine = '',
  language,
}: TextDisplayProps): JSX.Element {
  return (
    <>
      {
        text.lines.reduce<[JSX.Element[], Labels]>(
          (
            [elements, labels]: [JSX.Element[], Labels],
            line: AbstractLine,
            index: number
          ) => {
            const rows = skipLine(line, language)
              ? elements
              : [
                  ...elements,
                  <DisplayRow
                    key={index}
                    line={line}
                    lineIndex={index}
                    columns={text.numberOfColumns}
                    labels={labels}
                    activeLine={activeLine}
                    notes={text.notes}
                  />,
                ]
            return [rows, getCurrentLabels(labels, line)]
          },
          [[], defaultLabels]
        )[0]
      }
    </>
  )
}
function DisplayTwoColumnText({
  text,
  activeLine = '',
  language,
}: TextDisplayProps): JSX.Element {
  return (
    <>
      {
        text.lines.reduce<[JSX.Element[], Labels]>(
          (
            [elements, labels]: [JSX.Element[], Labels],
            line: AbstractLine,
            index: number
          ) => {
            const rulingsOffset = line.type === 'RulingDollarLine' ? 2 : 0
            const rows = isTranslationLine(line)
              ? elements
              : [
                  ...elements,
                  <DisplayRow
                    key={index}
                    line={line}
                    lineIndex={index}
                    columns={text.numberOfColumns + rulingsOffset}
                    labels={labels}
                    activeLine={activeLine}
                    notes={text.notes}
                  >
                    {language && (
                      <TranslationColumn
                        lines={text.lines}
                        lineIndex={index}
                        language={language}
                      />
                    )}
                  </DisplayRow>,
                ]
            return [rows, getCurrentLabels(labels, line)]
          },
          [[], defaultLabels]
        )[0]
      }
    </>
  )
}

export default function TransliterationLines({
  text,
  activeLine = '',
  translationStyle = 'inline',
  language = null,
}: {
  text: Text
  activeLine?: string
  translationStyle?: TranslationStyle
  language?: string | null
}): JSX.Element {
  const DisplayTextComponent =
    translationStyle === 'standoff'
      ? DisplayTwoColumnText
      : DisplaySingleColumnText
  return (
    <table className="Transliteration__lines">
      <tbody>
        <FirstLineNotes notes={text.notes} columns={text.numberOfColumns} />
        <DisplayTextComponent
          text={text}
          activeLine={activeLine}
          language={language}
        />
      </tbody>
    </table>
  )
}
