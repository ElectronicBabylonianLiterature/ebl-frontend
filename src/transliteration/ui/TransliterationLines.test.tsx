import React from 'react'
import _ from 'lodash'
import { render, within, RenderResult } from '@testing-library/react'
import noteLine from 'test-support/lines/note'
import { singleRuling } from 'test-support/lines/dollar'
import TransliterationLines from './TransliterationLines'
import { Text } from 'transliteration/domain/text'

const textWithNormalFirstLine = new Text({
  lines: [
    singleRuling,
    noteLine,
    noteLine,
    singleRuling,
    singleRuling,
    noteLine,
  ],
})
const normaFirsLineCases: [number, number[]][] = [
  [1, [1, 2]],
  [2, []],
  [3, [3]],
]

const textWithNoteOnFirstLine = new Text({
  lines: [noteLine, noteLine, singleRuling],
})
const noteFirstLineCases: [number, number[]][] = [
  [0, [1, 2]],
  [1, []],
]

let element: RenderResult
let lines: HTMLElement[]

describe.each([
  [textWithNormalFirstLine, false, normaFirsLineCases],
  [textWithNoteOnFirstLine, true, noteFirstLineCases],
])('%#', (text, noteOnFirstLine, lineCases: [number, number[]][]) => {
  beforeEach(() => {
    element = render(<TransliterationLines text={text} />)
    lines = element.getAllByRole('row')
  })

  test('Snapshot', () => {
    expect(element.container).toMatchSnapshot()
  })

  test('Shows all lines', () => {
    expect(lines.length).toEqual(lineCases.length)
  })

  describe.each(lineCases)(
    'Line %s with notes %p',
    (lineNumber, noteNumbers) => {
      let line: HTMLElement
      let links: HTMLElement[]

      beforeEach(() => {
        const index = noteOnFirstLine ? lineNumber : lineNumber - 1
        line = lines[index]
        links = within(line).queryAllByRole('link')
      })

      test('ID', () => {
        expect(line).toHaveAttribute('id', `line-${lineNumber}`)
      })

      test('Shows all links', () => {
        expect(links.length).toEqual(noteNumbers.length)
      })

      if (!_.isEmpty(noteNumbers)) {
        test.each(_.range(1, noteNumbers.length + 1))(
          '%s. link',
          (linkNumber) => {
            const index = linkNumber - 1
            const noteNumber = noteNumbers[index]
            const link = links[index]

            expect(link).toHaveAttribute('href', `#note-${noteNumber}`)
            expect(link).toHaveTextContent(String(noteNumber))
          }
        )
      }
    }
  )
})
