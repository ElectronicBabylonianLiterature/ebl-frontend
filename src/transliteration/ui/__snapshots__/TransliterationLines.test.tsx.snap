// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 Snapshot 1`] = `
<div>
  <table
    class="Transliteration__lines"
  >
    <tr
      id="line-1"
    >
      <td />
      <td
        class="Transliteration__RulingDollarLine"
        colspan="1"
      >
        <div
          class="Transliteration__ruling"
        />
      </td>
      <td>
         
        <a
          class="Transliteration__NoteLink"
          href="#note-1"
        >
          1
        </a>
         
        <a
          class="Transliteration__NoteLink"
          href="#note-2"
        >
          2
        </a>
         
      </td>
    </tr>
    <tr
      id="line-2"
    >
      <td />
      <td
        class="Transliteration__RulingDollarLine"
        colspan="1"
      >
        <div
          class="Transliteration__ruling"
        />
      </td>
      <td>
         
      </td>
    </tr>
    <tr
      id="line-3"
    >
      <td />
      <td
        class="Transliteration__RulingDollarLine"
        colspan="1"
      >
        <div
          class="Transliteration__ruling"
        />
      </td>
      <td>
         
        <a
          class="Transliteration__NoteLink"
          href="#note-3"
        >
          3
        </a>
         
      </td>
    </tr>
  </table>
</div>
`;

exports[`1 Snapshot 1`] = `
<div>
  <table
    class="Transliteration__lines"
  >
    <tr
      id="line-0"
    >
      <td />
      <td
        colspan="1"
      />
      <td>
         
        <a
          class="Transliteration__NoteLink"
          href="#note-1"
        >
          1
        </a>
         
        <a
          class="Transliteration__NoteLink"
          href="#note-2"
        >
          2
        </a>
         
      </td>
    </tr>
    <tr
      id="line-1"
    >
      <td />
      <td
        class="Transliteration__RulingDollarLine"
        colspan="1"
      >
        <div
          class="Transliteration__ruling"
        />
      </td>
      <td>
         
      </td>
    </tr>
  </table>
</div>
`;
